[Share 2.0] Move interfaces to OCP

rullzer opened the pull request:
As promised @DeepDiver1975 

It is time for the new sharing code to get out of the shadows and into the light!

CC: @PVince81 @DeepDiver1975 @nickvergessen @schiesbn @MorrisJobke @LukasReschke

rullzer committed:
[Share 2.0] Fix IShare

rullzer committed:
[Share 2.0] Move IShare to OCP

rullzer committed:
[Share 2.0] Create IManger interface for share manager

rullzer committed:
[Share 2.0] Add getShareManager to OCP\IServerContainer

MorrisJobke commented:
Shares still work and the change looks good. Also there are no errors in the log :+1:

PVince81 commented on lib/public/share/imanager.php:
```diff
@@ -0,0 +1,210 @@
+	 * @return IShare
+	 * @since 9.0.0
+	 */
+	public function newShare();
```
how is this different from `createShare` ? (make the docs clear enough).

I get you first need to instantiate a share (not really creating), then set its attributes, the pass it to `createShare`.

PVince81 commented on lib/public/share/imanager.php:
```diff
@@ -0,0 +1,210 @@
+	 * @return bool
+	 * @since 9.0.0
+	 */
+	public function isSharingDisabledForUser(IUser $user);
```
Inconsistent use of `is`.
Either remote `is` prefix here or add it everywhere above like `isShareApiLinkDefaultExpireDate` or `hasShareApiLinkDefaultExpireDate`

PVince81 commented on lib/public/share/ishare.php:
```diff
@@ -39,7 +47,8 @@ public function getId();
+	 * @return \OCP\Share\IShare The modified share object
+	 * @since 9.0.0
 	 */
 	public function setId($id);
```
Is an API consumer supposed to use this method at all ?

rullzer replied:
No in principal... but else unit testing becomes a lot harder... since then all objects have to be mocked... which is a bit of a pain sometimes...

rullzer replied:
SO that is not a very good argument... but sometimes using the actual object just is nicer... of course we could add a helper function somewhere in the test framework that just returns an internal object instead of the IShare... that would solve it as well..

rullzer replied:
Ok lets remove it for now... should not hurt... besides some warnings in the code maybe...

PVince81 commented on lib/public/share/ishare.php:
```diff
@@ -47,6 +56,7 @@ public function setId($id);
 	 * @return string
+	 * @since 9.0.0
 	 */
 	public function getFullId();
```
What's a full share id vs share id ? (explain in the comment)

PVince81 commented on lib/public/share/ishare.php:
```diff
@@ -54,7 +64,8 @@ public function getFullId();
+	 * @return \OCP\Share\IShare The modified share object\
+	 * @since 9.0.0
 	 */
 	public function setProviderId($id);
```
is an API consumer supposed to be able to set this ? (or is this automatic when retrieveing shares)

PVince81 commented on lib/public/share/ishare.php:
```diff
@@ -62,14 +73,16 @@ public function setProviderId($id);
+	 * @return \OCP\Share\IShare The modified object
+	 * @since 9.0.0
 	 */
 	public function setPath(Node $path);
```
=> "the path of the file/folder that was shared" ?

PVince81 commented on lib/public/share/ishare.php:
```diff
@@ -62,14 +73,16 @@ public function setProviderId($id);
 	public function setPath(Node $path);
 
 	/**
 	 * Get the path of this share for the current user
```
Not the same description as for `setPath` ?

You might want to change the text to "Returns the node representing the path" or something

PVince81 commented on lib/public/share/ishare.php:
```diff
@@ -100,6 +116,7 @@ public function setSharedWith($sharedWith);
 	 * Get the receiver of this share
 	 *
 	 * @return IUser|IGroup|string
```
when is this a string ? (add to the doc)

rullzer replied:
yeah this was future thinking for federated shares. But lets remove it for now. So we only add it when we actually use it.

PVince81 commented on lib/public/share/ishare.php:
```diff
@@ -137,7 +158,8 @@ public function getExpirationDate();
 	 * Set the sharer of the path
 	 *
 	 * @param IUser|string $sharedBy
```
when is this a string ? user name or is it specific to fed shares ? (explain in the PHPdoc)

PVince81 commented on lib/public/share/ishare.php:
```diff
@@ -152,15 +175,16 @@ public function getSharedBy();
 	 * Set the original share owner (who owns the path)
 	 *
 	 * @param IUser|string
```
when is this a string?

PVince81 commented on lib/public/share/ishare.php:
```diff
@@ -152,15 +175,16 @@ public function getSharedBy();
 	 * Get the original share owner (who owns the path)
-	 * 
+	 *
 	 * @return IUser|string
```
when is this a string?

PVince81 commented on lib/public/share/ishare.php:
```diff
@@ -177,6 +201,7 @@ public function setPassword($password);
 	 * Is a password set for this share
```
=> "Returns the password" ? It returns a string

PVince81 commented on lib/public/share/ishare.php:
```diff
@@ -184,7 +209,8 @@ public function getPassword();
 	 * Set the token
```
=> set the public link token

PVince81 commented on lib/public/share/ishare.php:
```diff
@@ -199,6 +226,7 @@ public function getToken();
 	 * Get the parent it
```
the parent what?

Is that the id of the parent share ? (is there still such a thing, I remember you said it was unused ?)

rullzer replied:
yeah needs to go! Thanks for spotting

PVince81 commented on lib/public/share/ishare.php:
```diff
@@ -206,7 +234,8 @@ public function getParent();
 	 * Set the target of this share
 	 *
 	 * @param string $target
```
the target path ? relative to what ?

PVince81 commented on lib/public/share/ishare.php:
```diff
@@ -214,20 +243,41 @@ public function setTarget($target);
 	 * Get the target of this share
```
target path?

PVince81 commented on lib/public/share/ishare.php:
```diff
@@ -214,20 +243,41 @@ public function setTarget($target);
 	/**
+	 * Set the time this share was created
+	 *
+	 * @param int $shareTime
```
Why not DateTime as with expiration, for consistency ? (it's fine to keep it as is)

rullzer replied:
Yeah that is cleaner. Fixed

PVince81 commented on lib/public/share/ishare.php:
```diff
@@ -214,20 +243,41 @@ public function setTarget($target);
+	public function getShareTime();
+
+	/**
+	 * Set mailSend
```
what is "mailSend" ?

rullzer replied:
fixed.

PVince81 commented on lib/public/share/ishareprovider.php:
```diff
@@ -18,53 +18,64 @@
 	public function identifier();
 
 	/**
 	 * Share a path
```
Sounds more like "Create the given share" or even "Persist/save the given share".
The fact that it's a path is implicit in the IShare object.

rullzer replied:
fixed

PVince81 commented on lib/public/share/ishareprovider.php:
```diff
@@ -18,53 +18,64 @@
 
 	/**
 	 * Unshare a file from self as recipient.
 	 * This may require special handling.
```
what kind of special handling ?

rullzer replied:
fixed

rullzer committed:
[Share 2.0] Fix interfaces and comments

* Made comments more clear
* Removed unneeded methods
* IShares shareTime is now a proper DateTime object
* IShares getPath -> getNode & setPath -> setNode
* Fix unit tests

rullzer commented:
@PVince81 addressed most of the comments.

* Cleaned up the interfaces.
* IShare setPath and getPath => setNode and getNode.
  - This is just cleaner since it is actually a node. And avoids things like $share->getPath()->getPath()
  - @DeepDiver1975 this might be relevant for your code. (might break some stuf). But I think this is a lot nicer.

rullzer committed:
[Share 2.0] Remove setId and setProviderId from the interface

Those should only ever be set by the providers

PVince81 commented:
Looks better.

Now I'm wondering what happens if I call `createShare` with a share that was already persisted.
Will it detect it and save it still or will it throw an exception ?
I see that only `InvalidShare` was published as exception in `@throws`.
We might want to add more official public exceptions like `ShareAlreadyExistsException` or `ShareNotFoundException` (or something more generic).

I'm fine with doing this later, can you add to your todo list ?

PVince81 commented:
:+1:

MorrisJobke commented:
Looks good :+1:

MorrisJobke commented:
> Now I'm wondering what happens if I call createShare with a share that was already persisted.
> Will it detect it and save it still or will it throw an exception ?
> I see that only InvalidShare was published as exception in @throws.
> We might want to add more official public exceptions like ShareAlreadyExistsException or ShareNotFoundException (or something more generic).

@PVince81 @rullzer I added it to the checklist in https://github.com/owncloud/core/issues/19331
